### PR TITLE
refactor: create general package structs

### PR DESCRIPTION
**Pull request by dsherret**
This is a bit of a balance between code re-use and not accidentally mixing different types of package structs. I tested this downstream in all the crates.

### CODE REVIEW

File: src/package.rs [hunk 1]
```diff
@@ -0,0 +1,519 @@
+use std::cmp::Ordering;
```
Comment (bartlomieju) [inline src/package.rs:1]: Add copyright?

File: src/package.rs [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+      }
+    })
+  }
+
+  pub(crate) fn as_specifier(&self, kind: PackageKind) -> Url {
+    let version_text = self.nv.version.to_string();
+    let scheme = kind.scheme();
+    let capacity = scheme.len() + 2 /* colon slash */
+      + self.nv.name.len()
+      + 1 /* @ */
```
Comment (bartlomieju) [inline src/package.rs:349]: Is this correct if the version is "*"?

**Review by bartlomieju — APPROVED** [review]
LGTM, nice cleanup

File: src/package.rs (hunk 2 above)
Comment (dsherret) [reply src/package.rs:349]: The version will always be resolved here, so it will be something like x.x.x

**Pull request merged**